src/main/java/com/mongodb/hibernate/service/MongoClientCustomizer.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.service;
+
+import com.mongodb.ConnectionString;
+import com.mongodb.MongoClientSettings;
+import org.hibernate.cfg.JdbcSettings;
+import org.hibernate.service.Service;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * A Hibernate {@link Service} focusing on customizing {@link com.mongodb.client.MongoClient} creation programmatically
+ * by invoking the various methods of {@link MongoClientSettings.Builder} before its
+ * {@link MongoClientSettings.Builder#build()} invocation.
+ *
+ * <p>An example usage is as follows:
+ *
+ * <pre>{@code
+ * var cfg = new Configuration();
+ *
+ * // configure cfg as you normally do (e.g. add entity classes)
+ * ...
+ *
+ * var clientCustomizer = (clientSettingsBuilder, connectionString) -> {
+ *     // customize client settings to your heart's content
+ *     ...
+ * };
+ *
+ * var serviceRegistryBuilder = cfg.getStandardServiceRegistryBuilder();
+ * serviceRegistryBuilder.addService(MongoClientCustomizer.class, clientCustomizer);
+ *
+ * var sessionFactory = cfg.buildSessionFactory();
+ *
+ * // start using sessionFactory as normal
+ * ...
+ * }</pre>
+ */
+@FunctionalInterface
+public interface MongoClientCustomizer extends Service {
+
+    /**
+     * Customize {@link MongoClientSettings} building.
+     *
+     * @param builder a {@link MongoClientSettings.Builder} instance which has been created with blank state (without
+     *     applying the connectionString parameter if provided)
+     * @param connectionString the {@link ConnectionString} created from {@value JdbcSettings#JAKARTA_JDBC_URL}
+     *     configuration if provided; provided for reference alone; could be null
```

Comment:
The documentation of Hibernate ORM uses multiple terms to refer to "configuration property" (the thing that has name, e.g., `jakarta.persistence.jdbc.url`, and value):

- "configuration setting", see https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#settings
- just "setting", see _"The settings ..."_ in https://docs.jboss.org/hibernate/orm/6.6/javadocs/org/hibernate/cfg/AvailableSettings.html.
- "configuration property", https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#envers-configuration
- just "property", see the `property` XML element in `hibernate.cfg.xml`.

Such pointless term overloading is not good, but then on top of that, our Hibernate dialect refers to those things as just "configuration"<sup>1</sup>. I propose us to pick a single term of the aforementioned (preferably, "configuration property", as it does not conflict with `MongoClientSettings`), and stick to it, until we have too large of a codebase to fix.

<sup>1</sup> Using the term "configuration" to refer to a single configuration property is especially bad because it conflicts with `org.hibernate.cfg.Configuration`, and because "configuration" is a collective term, not usually used to refer to a single configuration property.

Reply:
Good point! Let us use `configuration property` then. Updated. I also struggled whether to use `configValues` or `configurationValues` for the `Map<String, Object>` parameter in `MongoConnectionProvider#configure()`. Seems `configurationValues` are more common, but `configValues` shows up as well.

Reply:
There are multiple other places where this change applies to, but they were left unchanged.